Scalar: clean up the authentication story

dscho opened the pull request:
It is a bit tricky to align the C version of Scalar with the .NET version when it comes to authentication: the .NET version took care of it explicitly, and contains some logic to avoid authenticating _just_ for the `vsts/info` endpoint of a remote repository when the `gvfs/config` endpoint did not require authentication.

This matters because Scalar's functional tests depend on that behavior.

As a work-around, I already tried to clamp down any authentication apart from credential helper-based one, when running in unattended mode. But the functional tests do not run in unattended mode.

So this PR is an attempt to rectify this.

The idea is to prevent terminal-based authentication altogether, and if interactive credential usage was denied by the user (via `credential.interactive=never`), we now pretend that the user entered an empty password (via `GIT_ASKPASS=true`).

mjcheetham commented on contrib/scalar/scalar.c:
```diff
@@ -68,6 +76,24 @@ static void setup_enlistment_directory(int argc, const char **argv,
 	setup_git_directory();
+
+	/* Scalar requires the use of a credential helper */
+	setenv("GIT_TERMINAL_PROMPT", "false", 0);
```
This seems wrong.. we're never going to allow terminal-based prompting (including in GCM).
GCM checks `GIT_TERMINAL_PROMPT` too!

mjcheetham commented on contrib/scalar/scalar.c:
```diff
@@ -68,6 +76,24 @@ static void setup_enlistment_directory(int argc, const char **argv,
 	setup_git_directory();
+
+	/* Scalar requires the use of a credential helper */
+	setenv("GIT_TERMINAL_PROMPT", "false", 0);
```
This seems wrong.. we're never going to allow terminal-based prompting (including in GCM).
GCM checks `GIT_TERMINAL_PROMPT` too!

dscho replied:
What I meant to say is that we want to disable Git's own, terminal-based prompt, in favor of credential helpers. Does that sound okay?

derrickstolee replied:
Maybe this is something we should be setting in the functional tests.

dscho replied:
Heh, I just had the same idea in https://github.com/microsoft/git/pull/341#discussion_r631008550

mjcheetham replied:
By setting `GIT_TERMINAL_PROMPT=false` you are also disabling GCM however.

derrickstolee replied:
Our functional tests only test against public repos, so that shouldn't be  problem? I don't think GCM Core is even installed for the Linux and macOS builds.

mjcheetham commented on contrib/scalar/scalar.c:
```diff
@@ -68,6 +76,24 @@ static void setup_enlistment_directory(int argc, const char **argv,
+	/* Scalar requires the use of a credential helper */
+	setenv("GIT_TERMINAL_PROMPT", "false", 0);
+	if (is_unattended())
+		git_config_push_parameter("credential.interactive=never");
```
Just as an FYI `credential.interactive` is specific to GCM4W and GCMC.. other helpers do not care for this setting.
Therefore, there is an implicit assumption the configured helper for Scalar is GCM. Might want to call this out, or make it a piece of work to ensure this assumption is removed.

dscho replied:
Oh, right...

Hmm. I really do not know what else we could do to prevent that authenticated `vsts/info` endpoint from _at least_ producing ugly "error: config: (http:401) Not Authorized" lines.

I guess we could special-case Scalar's functional tests, by setting an environment variable that will skip looking at that endpoint. Probably what we should do. Do you concur?

mjcheetham reviewed (commented):
Setting `GIT_TERMINAL_PROMPT=false` in all cases seems wrong here.

dscho committed:
scalar: add a test toggle to skip accessing the vsts/info endpoint

In Scalar's functional tests, we do not do anything with authentication.
Therefore, we do want to avoid accessing the `vsts/info` endpoint
because it requires authentication even on otherwise public
repositories.

Let's introduce the environment variable `SCALAR_TEST_SKIP_VSTS_INFO`
which can be set to `true` to simply skip that step (and force the
`url_*` style repository IDs instead of `id_*` whenever possible).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: implement 'scalar list'

We should not fail to unregister the enlistment just because the
FSMonitor could not be stopped (e.g. because the directory was already
removed behind Scalar's back).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
Okay, I am doing things very differently now: `SCALAR_TEST_SKIP_VSTS_INFO=true` skips that `vsts/info` endpoint, and I am only waiting for https://dev.azure.com/gvfs/ci/_build/results?buildId=19757&view=logs&j=00af5505-930d-50dd-3844-b167e8b4bb13&t=7ab88860-f9c4-5cd4-8cf2-8c804abec61b to finish before I push an update to https://github.com/microsoft/scalar/pull/505 that sets this environment variable.
